service/popm/popm.go

```diff
@@ -50,6 +50,112 @@ func init() {
 	loggo.ConfigureLoggers(logLevel)
 }
 
+// L2KeystonePriorityBuffer holds up to "size" L2Keystones. it allows a caller
+// to push to it and it will keep the most recent "size" keystones, overriding
+// the oldest if full.  we define "oldest" as "the smallest l2 block number"
+type L2KeystonePriorityBuffer struct {
+	mtx     sync.Mutex
+	mapping map[string]*L2KeystonePriorityBufferElement
+	size    int
+}
+
+// L2KeystonePriorityBufferElement holds an L2Keystone and whether it has been
+// "processed" or not
+type L2KeystonePriorityBufferElement struct {
+	l2Keystone         hemi.L2Keystone
+	requiresProcessing bool
+}
+
+// NewL2KeystonePriorityBuffer creates a L2KeystonePriorityBuffer with size n
+func NewL2KeystonePriorityBuffer(n int) *L2KeystonePriorityBuffer {
+	return &L2KeystonePriorityBuffer{
+		mapping: make(map[string]*L2KeystonePriorityBufferElement),
+		size:    n,
+	}
+}
+
+// Push inserts an L2Keystone, dropping the oldest if full
+func (r *L2KeystonePriorityBuffer) Push(val hemi.L2Keystone) {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+
+	item := L2KeystonePriorityBufferElement{
+		l2Keystone:         val,
+		requiresProcessing: true,
+	}
+
+	serialized := hemi.L2KeystoneAbbreviate(val).Serialize()
+	key := hex.EncodeToString(serialized[:])
+
+	// keystone already exists, no-op
+	if _, ok := r.mapping[key]; ok {
+		return
+	}
+
+	if len(r.mapping) < r.size {
+		r.mapping[key] = &item
+		return
+	}
+
+	var smallestL2BlockNumber uint32
+	var smallestKey string
+
+	for k, v := range r.mapping {
+		if smallestL2BlockNumber == 0 || v.l2Keystone.L2BlockNumber < smallestL2BlockNumber {
+			smallestL2BlockNumber = v.l2Keystone.L2BlockNumber
+			smallestKey = k
+		}
+	}
+
+	// do not insert an L2Keystone that is older than all of the ones already
+	// queued
+	if item.l2Keystone.L2BlockNumber < smallestL2BlockNumber {
+		return
+	}
+
+	delete(r.mapping, smallestKey)
+
+	r.mapping[key] = &item
+}
+
+// ForEach is a thread-safe function that calls a callback function to "process"
+// each L2Keystone.  The callback function is called with a copy of the
+// L2Keystone to process.  if the callback returns an error, no-op, otherwise
+// mark the L2Keystone as processed
+func (r *L2KeystonePriorityBuffer) ForEach(cb func(ks hemi.L2Keystone) error) {
+	r.mtx.Lock()
+	copies := []hemi.L2Keystone{}
+	for _, v := range r.mapping {
+		if v.requiresProcessing {
+			copies = append(copies, v.l2Keystone)
+
+			// temporarily set this to false, so another goroutine doesn't
+			// try to process
+			v.requiresProcessing = false
+		}
+	}
+	r.mtx.Unlock()
+
+	// mine the newest keystone first
+	slices.SortFunc(copies, func(a, b hemi.L2Keystone) int {
+		return int(b.L2BlockNumber) - int(a.L2BlockNumber)
+	})
+
+	for _, e := range copies {
+		mined := hemi.L2KeystoneAbbreviate(e).Serialize()
+		key := hex.EncodeToString(mined[:])
+
+		err := cb(e)
+		r.mtx.Lock()
```

Comment:
I didn't want to say this but this is probably where you want to serialize using a channel. I may be missing some subtlety so don't take this as gospel truth but once you start seeing locking/unlocking in the same loop it typically means the code should be rethought to prevent that from happening.